sa/sa.go

```diff
@@ -2115,8 +2113,39 @@ func (ssa *SQLStorageAuthority) GetAuthorizations2(ctx context.Context, req *sap
 		return nil, err
 	}
 
+	authz2IDMap := map[int64]bool{}
+	// Once the old authorization storage format fallback is removed we don't need
+	// this length check as if there are none returned we can just return immediately.
+	if len(authzModels) > 0 {
+		// Instead of using a JOIN on the orderToAuthz2 table in order to filter out
+		// authorizations that don't have associated orders we do a second query
+		// and do the filtering in code. This allows us to bypass the large number
+		// of duplicate rows that are generated by using a JOIN.
+		var ids []interface{}
+		qmarks = make([]string, len(authzModels))
+		for i, am := range authzModels {
+			ids = append(ids, am.ID)
+			qmarks[i] = "?"
+		}
+		var authz2IDs []int64
+		_, err = ssa.dbMap.Select(
+			&authz2IDs,
+			fmt.Sprintf(`SELECT DISTINCT(authzID) FROM orderToAuthz2 WHERE authzID IN (%s)`, strings.Join(qmarks, ",")),
```

Comment:
Looking at this, it may turn out to be similarly expensive in the corner case we found, since it's potentially reading many `orderToAuthz2` rows for a single authorization that's included in many orders. Presumably it's still somewhat cheaper since at least it's not doing the join, so I think it's worthwhile, but not the big win we'd hoped for.

The only alternative I see would be to split this into N independent queries, each with a `LIMIT 1`, but that definitely increases the complexity and might actually make performance work. So on balance I think this approach is fine.